### DISCUSSION TITLE
Project MTP (mysterious cat adventure game) dev log

### DISCUSSION THREAD

**Discussion by eliasdaler** [eliasdaler/edw.is#6] [Article announcements / comments]
Finally started documenting my new 3D game dev progress here:

https://edw.is/project-mtp-devlog/

![image](https://github.com/eliasdaler/edw.is/assets/1285136/9efe728b-f860-45dc-9bfc-415952596883)

This is a continuation of this previous thread: https://github.com/eliasdaler/edw.is/discussions/4. So, for making "small" games, Lua is nice. But for bigger games (especially when doing non-trivial OpenGL rendering) C++ is important.

I'll post updates here as well so that people can comment and discuss it more easily. If you have any questions about the game or the engine, feel free to ask!

**Comment by eliasdaler** [comment]
## 2023-11-22 You dawg I heard you like games…

… so we put a game inside your game. This is an independent small game running inside of the main game. It’s rendered on the TV screen texture. 

I'll start working on a more non-trivial "inner" game soon as I want to see if I can reuse various ECS/UI/rendering components and have them not affect the main game. I could potentially just expose a very small API to the "inner" games (e.g. just "playSound" and "setPixel"), but that would mean that I'd have to a lot of the things twice (e.g. the dialogue box system, the input manager, NPC interaction etc.).

https://github.com/eliasdaler/edw.is/assets/1285136/7f469848-ec20-43da-8589-eece5d75ca35

**Comment by kafkaphoenix** [comment]
Hi, the game looks amazing so far! are you using imgui for your ui or only for the debugging editor?

**Reply by eliasdaler** [reply]
Thank you!

Yeah, ImGui is only used for the dev tools, I wrote in-game UI myself. Maybe some day I'll explain how it works, but basically it allows you to put all widgets into grid containers and then the system positions all the elements automatically, kinda like flexbox.

**Reply by kafkaphoenix** [reply]
Definitely interested in that future blog! It looks really impressive keep the hard work :)

**Comment by eliasdaler** [comment]
[On my website](https://edw.is/project-mtp-devlog-december-2023/)


https://github.com/eliasdaler/edw.is/assets/1285136/5ebfc61d-00d8-4cca-a0e8-886015694ab3

Currently working on resolving technical debt and porting my Ebitengine platformer game to new engine.

It’s going great so far!

--- 


https://github.com/eliasdaler/edw.is/assets/1285136/0cc1c723-7a23-438a-8738-a7ff16248830

(N64/PS1 switch, am I right?)

Started using sampler objects. They're fantastic. Previously I used what many people do in OpenGL tutorials, the sampling state was hardcoded into each texture like this:

```cpp
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
... // etc.
```

But sometimes you need to sample it differently. And different systems might want to sample the same texture in a different way (some with linear filtering, some with nearest). Having to manually set nearest filtering for UI textures was also annoying. 

Sampler objects to the rescue! Now I can just do this:

```cpp
const auto& sm = Engine::getShaderManager();
const auto& nearestSampler = sm.getDefaultNearestSampler();
const auto& shader = sm.getShader(shaderName);
shader.setSamplerUniform("tex", prevTarget.getTexture(), 0, nearestSampler);
```

or create a sampler like this:

```cpp
const auto& sampler =  sm.getSampler({
        .filter =
            {
                .min = edge::FilterType::Linear,
                .mag = edge::FilterType::Linear,
            },
        .wrap = SamplerObject::WrapParams::ClampToEdge,
        .anisotropy = 8.f,
    });
```
(all sampler objects get cached so that sampler object don't get re-created over and over)

**Comment by eliasdaler** [comment]
Closing for now, don't have much time to write frequent dev logs. :)
